pkg/connector/ble/ble.go

```diff
@@ -125,10 +125,89 @@ func (c *Connection) VIN() string {
 	return c.vin
 }
 
+func VehicleLocalName(vin string) string {
+	vinBytes := []byte(vin)
+	digest := sha1.Sum(vinBytes)
+	return fmt.Sprintf("S%02xC", digest[:8])
+}
+
+func initDevice() error {
+	var err error
+	// We don't want concurrent calls to NewConnection that would defeat
+	// the point of reusing the existing BLE device. Note that this is not
+	// an issue on MacOS, but multiple calls to newDevice() on Linux leads to failures.
+	if device != nil {
+		log.Debug("Reusing existing BLE device")
+	} else {
+		log.Debug("Creating new BLE device")
+		device, err = newDevice()
+		if err != nil {
+			return fmt.Errorf("failed to find a BLE device: %s", err)
+		}
+		ble.SetDefaultDevice(device)
+	}
+	return nil
+}
+
+type Advertisement = ble.Advertisement
+
+func ScanVehicleBeacon(ctx context.Context, vin string) (Advertisement, error) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if err := initDevice(); err != nil {
+		return nil, err
+	}
+
+	a, err := scanVehicleBeacon(ctx, VehicleLocalName(vin))
+	if err != nil {
+		return nil, fmt.Errorf("ble: failed to scan for %s: %s", vin, err)
+	}
+	return a, nil
+}
+
+func scanVehicleBeacon(ctx context.Context, localName string) (Advertisement, error) {
+	var err error
+	ctx2, cancel := context.WithCancel(ctx)
+	go func() {
+		select {
+		case <-ctx.Done():
+			cancel()
+		case <-ctx2.Done():
+		}
+	}()
+
+	ch := make(chan Advertisement)
+	fn := func(a Advertisement) {
+		ln := a.LocalName()
+		if ln != localName {
+			return
+		}
+		cancel()
+		ch <- a
+	}
+	err = device.Scan(ctx2, false, fn)
+
+	select {
+	case a, ok := <-ch:
+		if !ok {
+			// This should never happen, but just in case
+			return nil, fmt.Errorf("scan channel closed")
+		}
+		return a, nil
+	default:
+		return nil, err
+	}
+}
+
 func NewConnection(ctx context.Context, vin string) (*Connection, error) {
+	return NewConnectionToBleTarget(ctx, vin, nil)
+}
+
+func NewConnectionToBleTarget(ctx context.Context, vin string, target Advertisement) (*Connection, error) {
 	var lastError error
 	for {
-		conn, err := tryToConnect(ctx, vin)
+		conn, err := tryToConnect(ctx, vin, target)
```

Comment:
Results in redundant calls to `VehicleLocalName`. Call it once in `NewConnectionToBleTarget` and pass the result to `tryToConnect` instead of the VIN.

Reply:
I think the only way its called twice is when user is calling first `ScanVehicleBeacon` and then `NewConnectionToBleTarget`. If calling `NewConnection` (or `NewConnectionToBleTarget(ctx, vin, nil)`) it is only generated once in `tryToConnect` and passed to `scanVehicleBeacon`. I think avoiding double generation can be solved only by requiring the user to generate the local name and passing it to both the calls, replacing the vin argument in  both `ScanVehicleBeacon` and `TryToConnectToBleTarget`. Did you mean this?